libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,529 @@
+				// it further up once more.
+				// This would mean not stopping at _block == target above or even doing visited.clear() here, revisiting the entire graph.
+				// This is a tradeoff between the runtime of this process and the optimality of the result.
+				// Also note that while visiting the entire graph again *can* be helpful, it can also be detrimental.
+				// Also note that for some reason using visited.clear() is incompatible with skipping the revisit
+				// of already seen exit layouts above, I'm not sure yet why.
+			}
+	}
+
+	stitchConditionalJumps(_entry);
+	fixStackTooDeep(_entry);
+}
+
+Stack StackLayoutGenerator::combineStack(Stack const& _stack1, Stack const& _stack2)
+{
+	// TODO: there is probably a better way than brute-forcing. This has n! complexity or worse, so
+	// we can't keep it like this.
+
+	Stack commonPrefix;
+	for (auto&& [slot1, slot2]: ranges::zip_view(_stack1, _stack2))
+	{
+		if (!(slot1 == slot2))
+			break;
+		commonPrefix.emplace_back(slot1);
+	}
+
+	Stack stack1Tail = _stack1 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+	Stack stack2Tail = _stack2 | ranges::views::drop(commonPrefix.size()) | ranges::to<Stack>;
+
+	if (stack1Tail.empty())
+		return commonPrefix + compressStack(stack2Tail);
+	if (stack2Tail.empty())
+		return commonPrefix + compressStack(stack1Tail);
+
+	Stack candidate;
+	for (auto slot: stack1Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	for (auto slot: stack2Tail)
+		if (!util::findOffset(candidate, slot))
+			candidate.emplace_back(slot);
+	cxx20::erase_if(candidate, [](StackSlot const& slot) {
+		return holds_alternative<LiteralSlot>(slot) || holds_alternative<FunctionCallReturnLabelSlot>(slot);
+	});
+
+	std::map<size_t, Stack> sortedCandidates;
+
+	// TODO: surprisingly this works for rather comparably large candidate size, but we should probably
+	// set up some limit, since this will quickly explode otherwise.
+	// Ideally we would then have a better fallback mechanism - although returning any naive union of both stacks
+	// like ``candidate`` itself may just be fine.
+	//	if (candidate.size() > 8)
+	//		return candidate;
+
+	auto evaluate = [&](Stack const& _candidate) -> size_t {
+		size_t numOps = 0;
+		Stack testStack = _candidate;
+		auto swap = [&](unsigned _swapDepth) { ++numOps; if (_swapDepth > 16) numOps += 1000; };
+		auto dupOrPush = [&](StackSlot const& _slot)
+		{
+			if (canBeFreelyGenerated(_slot))
+				return;
+			auto depth = util::findOffset(ranges::concat_view(commonPrefix, testStack) | ranges::views::reverse, _slot);
+			if (depth && *depth >= 16)
+				numOps += 1000;
+		};
+		createStackLayout(testStack, stack1Tail, swap, dupOrPush, [&](){} );
+		testStack = _candidate;
+		createStackLayout(testStack, stack2Tail, swap, dupOrPush, [&](){});
+		return numOps;
+	};
+
+	// See https://en.wikipedia.org/wiki/Heap's_algorithm
```

Comment:
Can this use `std::next_permutation`?

Reply:
I quickly looked at it and concluded that it can't, but maybe I was too quick about that. Might even be better - the current algorithm is agnostic to equality of slots, but I'd need to check and generally this whole thing needs to be replaced by a direct constructive algorithm anyways. (For the record, it still contains this bug right now and doing it properly is actually too slow: https://github.com/ethereum/solidity/pull/11493#discussion_r657023427)